gobblin-config-management/src/main/java/gobblin/config/client/ConfigClient.java

```diff
@@ -0,0 +1,273 @@
+      value.resolver = resolver;
+    }
+
+    // create default importedBy mapping object
+    if (!((cs instanceof ConfigStoreWithImportedBy) && (cs instanceof ConfigStoreWithImportedByRecursively))) {
+      SimpleImportMappings im = null;
+      // need to create the SimpleImportMappings using the resolver NOT the raw configStore
+      // otherwise, the getImportsRecursively and getImportedByRecursively will not work
+      // as raw configstore is NOT ConfigStoreWithResolution
+      if (!(cs instanceof ConfigStoreWithResolution)) {
+        im = new SimpleImportMappings(value.resolver, value.version);
+      } else {
+        im = new SimpleImportMappings(cs, value.version);
+      }
+      value.simpleImportMappings = im;
+    }
+
+    this.configStoreMap.put(key, value);
+    LOG.info(String.format("Created new config store with uri: %s, version: %s", cs.getStoreURI(),
+        cs.getCurrentVersion()));
+    return value;
+  }
+
+  // use serviceLoader to load configStoreFactories
+  @SuppressWarnings("unchecked")
+  private ConfigStoreFactory<ConfigStore> getConfigStoreFactory(URI uri) throws Exception {
+    ConfigStoreFactory csf = configStoreFactoryMap.get(uri.getScheme());
+    if (csf == null) {
+      throw new Exception("can not find corresponding config store factory for scheme " + uri.getScheme());
+    }
+
+    return (ConfigStoreFactory<ConfigStore>) csf;
+  }
+
+  /**
+   * 
+   * @param uri - must start with scheme name
+   * @return - the directly and indirectly specified configuration in {@com.typesafe.config.Config} format for input uri 
+   * 
+   * <p>
+   * detail logics:
+   * 1. Used the ConfigStoreAccessor in the cached map if previous ConfigStore been queried
+   * 2. Based the scheme name, using {@java.util.ServiceLoader} to find the first {@gobblin.config.configstore.ConfigStoreFactory}
+   * 3. Use ConfigStoreFactory to create ConfigStore 
+   * 4. Build ConfigStoreAccessor by checking the current version of the ConfigStore. Added the entry to theMap
+   * 5. If the ConfigStore is NOT ConfigStoreWithResolution, need to do resolution in this client
+   */
+  public Config getConfig(URI uri) throws Exception {
+
+    ConfigStoreAccessor csa = this.getConfigStoreAccessor(uri);
+    URI rel_uri = csa.store.getStoreURI().relativize(uri);
+
+    if (csa.store instanceof ConfigStoreWithResolution) {
+      return ((ConfigStoreWithResolution) csa.store).getResolvedConfig(rel_uri, csa.version);
+    }
+
+    SimpleConfigStoreResolver resolver = csa.resolver;
+    return resolver.getResolvedConfig(rel_uri, csa.version);
+  }
+
+  /**
+   * @param uris - Collection of URI, each one much start with scheme name
+   * @return - the java.util.Map. Key of the map is the URI, value of the Map is getConfig(URI key)
+   */
+  public Map<URI, Config> getConfigs(Collection<URI> uris) throws Exception {
+    Map<URI, Config> result = new HashMap<URI, Config>();
```

Comment:
You can add a TODO comment that we can have a logic if the number URIs is large, we can parallelize the calls using the ThreadPool.